Projects/b_u585i_iot02a_ntz/Src/app_main.c

```diff
@@ -242,14 +242,72 @@ void vInitTask( void * pvArgs )
     }
 }
 
+static uint32_t ulCsrFlags = 0;
+
+static void vDetermineResetSource()
+{
+    const char * pcResetSource = NULL;
+
+    ulCsrFlags &= ( RCC_CSR_LPWRRSTF_Msk | RCC_CSR_WWDGRSTF_Msk |
+                    RCC_CSR_IWDGRSTF_Msk | RCC_CSR_SFTRSTF_Msk |
+                    RCC_CSR_BORRSTF_Msk | RCC_CSR_PINRSTF_Msk |
+                    RCC_CSR_OBLRSTF_Msk );
+
+    if( ulCsrFlags & RCC_CSR_LPWRRSTF_Msk )
+    {
+        pcResetSource = "LPWRRSTF";
+    }
+    else if( ulCsrFlags & RCC_CSR_WWDGRSTF_Msk )
+    {
+        pcResetSource = "WWDGRSTF: Windowing Watchdog";
+    }
+    else if( ulCsrFlags & RCC_CSR_IWDGRSTF_Msk )
+    {
+        pcResetSource = "IWDGRSTF: Independent Watchdog";
+    }
+    else if( ulCsrFlags & RCC_CSR_SFTRSTF_Msk )
+    {
+        pcResetSource = "SFTRSTF: Software";
+    }
+    else if( ulCsrFlags & RCC_CSR_BORRSTF_Msk )
+    {
+        pcResetSource = "BORRSTF: Brownout";
+    }
+    else if( ulCsrFlags & RCC_CSR_PINRSTF_Msk )
+    {
+        pcResetSource = "PINRSTF: NSRT Pin";
+    }
+    else if( ulCsrFlags & RCC_CSR_OBLRSTF_Msk )
+    {
+        pcResetSource = "OBLRSTF: Option Byte Load";
+    }
+    else
+    {
+        pcResetSource = "Unknown";
+    }
+
+    LogSys( "Reset Source: 0x%x : %s.", ulCsrFlags, pcResetSource );
+}
+
+uint32_t ulGetResetSource( void )
+{
+    return ulCsrFlags;
+}
+
 int main( void )
 {
+    ulCsrFlags = RCC->CSR;
+
+    __HAL_RCC_CLEAR_RESET_FLAGS();
```

Comment:
Do we need this for TFM code?

Reply:
I've only tested the watchdog portion on the TFM version. The RCC CSR flags are probably cleared somewhere in the bootloader or TFM.. but might not be.